### PR TITLE
Custom validation and test for complex_filter dsl

### PR DESCRIPTION
**Pull request by stewartfrancis**

### CODE REVIEW

File: plugins/module_utils/cmci.py [hunk 1]
```diff
@@ -408,6 +388,7 @@ def init_url(self):  # type: () -> str
     def init_request_params(self):  # type: () -> Optional[Dict[str, str]]
         return None
 
+    # TODO: why do we have this and/or parsing twice?
```
Comment (sophiegreen) [inline plugins/module_utils/cmci.py:391]: Did you want to raise an issue for this and remove the TODO?

File: plugins/module_utils/cmci.py [hunk 2]
```diff
@@ -603,67 +587,153 @@ def append_attributes(self, element):
...
+            diff = set(i.keys()) - set(valid_keys)
+            if len(diff) != 0:
+                self._fail(
+                    "Unsupported parameters for (basic.py) module: %s found"
+                    " in resources -> complex_filter%s. Supported parameters "
+                    "include: %s"
+                    % (", ".join(diff), path, ", ".join(valid_keys))
+                )
 
+            # Validate required_one_of
```
Comment (sophiegreen) [inline plugins/module_utils/cmci.py:616]: Does this validation still need adding or is it done?

Comment (sophiegreen) [reply plugins/module_utils/cmci.py:616]: Scrap that, found it below

File: plugins/module_utils/cmci.py [hunk 3]
```diff
@@ -603,67 +587,153 @@ def append_attributes(self, element):
...
 
-            complex_filter_string = _append_filter_string(
-                complex_filter_string,
-                attribute_filter_string,
-                joiner
-            )
+            # Validate required_by
+            op = i.get(OPERATOR)
+
+            if op and not attribute:
```
Comment (sophiegreen) [inline plugins/module_utils/cmci.py:631]: Do we need to do the same for a value with no attribute?

File: tests/sanity/ignore-2.10.txt [hunk 4]
```diff
@@ -1,38 +1,6 @@
...
-plugins/modules/cmci_update.py validate-modules:undocumented-parameter # We have a complex_filter DSL as part of our plugin YAML.  The same config is accepted 4 levels deep.  We only include the (quite long) documentation for this at the highest level, which causes this error
-plugins/modules/cmci_update.py validate-modules:doc-elements-mismatch # We have a complex_filter DSL as part of our plugin YAML.  The same config is accepted 4 levels deep.  We only include the (quite long) documentation for this at the highest level, which causes this error
-plugins/modules/cmci_update.py validate-modules:mutually_exclusive-unknown # We have a complex_filter DSL as part of our plugin YAML.  The same config is accepted 4 levels deep.  We only include the (quite long) documentation for this at the highest level, which causes this error
-plugins/modules/cmci_update.py validate-modules:required_one_of-unknown # We have a complex_filter DSL as part of our plugin YAML.  The same config is accepted 4 levels deep.  We only include the (quite long) documentation for this at the highest level, which causes this error
```
Comment (sophiegreen) [inline tests/sanity/ignore-2.10.txt:38]: 🥳

File: plugins/module_utils/cmci.py (hunk 1 above)
Comment (stewartfrancis) [reply plugins/module_utils/cmci.py:391]: I was actually gonna try and fix it, forgot I'd left the TODO in there... let me see if I can use the same stuff twice!

File: plugins/module_utils/cmci.py (hunk 3 above)
Comment (stewartfrancis) [reply plugins/module_utils/cmci.py:631]: Yes, I think you're right... let me write a test and see if something weird happens.  My go to...

**Review by sophiegreen — APPROVED** [review]

**Pull request merged**